File: Systems/MediaPlayerSystem.py
import codecs
import logging
import os
import socket
from configparser import RawConfigParser
from datetime import datetime, timedelta

import vlc
from nmb.NetBIOS import NetBIOS
from smb.SMBConnection import SMBConnection

from Systems.BaseSystem import BaseSystem
from Utils.Decorators import try_catch, type_check

logger = logging.getLogger(__name__.split(".")[-1])


class MediaPlayerSystem(BaseSystem):
    """ MediaPlayerSystem class """
    SupportedFormats = [".mkv", ".avi", ".mov", ".wmv",
                        ".mp4", ".mpg", ".mpeg", ".m4v", ".3gp", ".mp3"]

    @type_check
    def __init__(self, owner: None) -> None:
        """ Initialize an instance of the MediaPlayerSystem class.

        Arguments:
                owner {MyHome} -- MyHome object which is the owner of the system.
        """

        super().__init__(owner)
        self._isEnabled = None

        self.mediaPaths = ["."]
        self.sharedPath = ""
        self.volume = 0
        self.radios = []

        self._timer = datetime.now() - timedelta(minutes=1)
        self._player = vlc.MediaListPlayer()
        self._sharedList = []
        self._playing = ""
        self._watched = []

    @type_check
    def load(self, configParser: RawConfigParser, data: dict) -> None:
        """ Loads settings and data for the system.

        Arguments:
                configParser {RawConfigParser} -- ConfigParser from which the settings will be loaded.
                data {dict} -- Dictionary from which the system data will be loaded.
        """

        super().load(configParser, data)

        if "sharedList" in data:
            self._sharedList = data["sharedList"]
        if "watched" in data:
            self._watched = data["watched"]

    @type_check
    def save(self, configParser: RawConfigParser, data: dict) -> None:
        """ Saves settings and data used by the system.

        Arguments:
                configParser {RawConfigParser} -- ConfigParser to which the settings will be saved.
                data {dict} -- Dictionary to which the system data will be saved.
        """

        super().save(configParser, data)

        data["sharedList"] = self._sharedList
        data["watched"] = self._watched

    @type_check
    def update(self) -> None:
        """ Update current system's state. """

        super().update()

        if datetime.now() - self._timer < timedelta(minutes=1):
            return
        self._timer = datetime.now()

        # stop if end reached
        if self._player.get_state() == vlc.State.Ended:
            self._player.stop()

    @property
    @type_check
    def mediaList(self) -> list:
        """ Gets a list with all playable media.

        Returns:
            list -- List of media files.
        """

        result = []
        for i, mediaPath in enumerate(self.mediaPaths):
            if os.path.isdir(mediaPath):
                for root, subFolders, files in os.walk(mediaPath):
                    subFolders.sort()
                    files.sort()
                    for f in files:
                        if os.path.splitext(f)[1] in self.SupportedFormats:
                            path = os.path.join(root, f)
                            prefix = f"local{i+1}" + os.path.sep
                            result.append(
                                prefix + os.path.relpath(path, mediaPath))
        # shared
        result.extend(["shared" + os.path.sep + i for i in self._sharedList])
        # add radios
        result.extend(["radios" + os.path.sep + i for i in self.radios])
        return result

    @property
    @type_check
    def mediaTree(self) -> dict:
        """ Gets a tree view with all playable media.

        Returns:
            dict -- Dictionary with all playable media.
        """

        tree = {}
        for f in self.mediaList:
            split = f.split(os.path.sep)
            currItem = tree
            for s in split:
                if s not in currItem:
                    currItem[s] = {}
                currItem = currItem[s]
        return tree

    @property
    @type_check
    def playing(self) -> str:
        """ Gets the playing file.

        Returns:
            str -- The playing file.
        """

        if self._player.get_state() == vlc.State.Stopped:
            self._playing = ""
        return self._playing

    @property
    @type_check
    def timeDetails(self) -> str:
        """ Gets timing details (current time / length).

        Returns:
            str -- The timing details (current time / length).
        """
        def getTime(ms):
            s = ms / 1000
            m, s = divmod(s, 60)
            h, m = divmod(m, 60)
            return int(h), int(m), int(s)

        time = getTime(self._player.get_media_player().get_time())
        length = getTime(self._player.get_media_player().get_length())
        return f"{time[0]}:{time[1]:02} / {length[0]}:{length[1]:02}"

    @type_check
    def play(self, path: str) -> None:
        """ Play the set media file.

        Arguments:
            path {str} -- Media file to be played.
        """

        logger.debug("Play media: %s", path)

        if path == "":
            return

        self._playing = path
        if not path.startswith("http") and not path.startswith("https"):  # URL
            _type = path[:path.index(os.path.sep)]
            # remove local/shared/radios prefix
            path = path[path.index(os.path.sep) + 1:]
        else:
            _type = "radios"
        if _type == "radios":
            # TODO: set local audio output
            print(self._player.get_instance().audio_output_enumerate_devices())
            self._player.set_media_list(vlc.MediaList([path]))
            self._player.get_media_player().audio_set_volume(
                int(100 + self.volume * 5 * 1.5))
        else:
            if _type.startswith("local"):
                mediaPath = self.mediaPaths[int(_type[5:]) - 1]
                path = os.path.join(mediaPath, path)
                self._convertSubtitles(path)
            else:
                path = self.sharedPath + path
            self._player.set_media_list(vlc.MediaList([path]))
            self._player.get_media_player().audio_set_volume(int(100 + self.volume * 5))

        self._player.get_media_player().set_fullscreen(True)
        self._player.play()
        if _type != "radios":
            self._markAsWatched(self._playing)
        self._owner.event(self, "MediaPlayed", self._playing)

    @type_check
    def stop(self) -> None:
        """ Stop the current playing. """

        logger.debug("Stop media: %s", self._playing)
        self._playing = ""
        self._player.stop()
        self._owner.event(self, "MediaStopped")

    @type_check
    def pause(self) -> None:
        """ Pause the current playing. """

        logger.debug("Pause media: %s", self._playing)
        self._player.pause()
        self._owner.event(self, "MediaPaused")

    @type_check
    def volumeDown(self) -> None:
        """ Volume down the current playing. """

        logger.debug("Volume down media: %s", self.volume)
        self._player.get_media_player().audio_set_volume(
            self._player.get_media_player().audio_get_volume() - 5)
        self.volume -= 1
        self._owner.systemChanged = True
        self._owner.event(self, "MediaVolumeDown")

    @type_check
    def volumeUp(self) -> None:
        """ Volume up the current playing. """

        logger.debug("Volume up media: %s", self.volume)
        self._player.get_media_player().audio_set_volume(
            self._player.get_media_player().audio_get_volume() + 5)
        self.volume += 1
        self._owner.systemChanged = True
        self._owner.event(self, "MediaVolumeUp")

    @type_check
    def seekBack(self) -> None:
        """ Seek back the current playing. """

        logger.debug("Seek back media")
        self._player.get_media_player().set_time(
            self._player.get_media_player().get_time() - 30 * 1000)  # -30 seconds
        self._owner.event(self, "MediaSeekBack")

    @type_check
    def seekForward(self) -> None:
        """ Seek forward the current playing. """

        logger.debug("Seek forward media")
        self._player.get_media_player().set_time(
            self._player.get_media_player().get_time() + 30 * 1000)  # +30 seconds
        self._owner.event(self, "MediaSeekForward")

    @type_check
    def seekBackFast(self) -> None:
        """ Seek back fast the current playing. """

        logger.debug("Seek back fast media")
        self._player.get_media_player().set_time(
            self._player.get_media_player().get_time() - 600 * 1000)  # -600 seconds
        self._owner.event(self, "MediaSeekBackFast")

    @type_check
    def seekForwardFast(self) -> None:
        """ Seek forward fast the current playing. """

        logger.debug("Seek forward fast media")
        self._player.get_media_player().set_time(
            self._player.get_media_player().get_time() + 600 * 1000)  # +600 seconds
        self._owner.event(self, "MediaSeekBackFast")

    @type_check
    def refreshSharedList(self) -> None:
        """ Refresh the list with shared files. """

        logger.debug("Refresh shared list")
        self._sharedList = self._listShared()
        self._owner.systemChanged = True

    @type_check
    def _markAsWatched(self, path: str) -> None:
        """ Mark the set media path as watched.

        Arguments:
            path {str} -- Path to be set as watched.
        """

        logger.debug("Mark as watched: %s", path)
        # cleanup watched list from nonexistent files
        self._watched = [w for w in self._watched if w in self.mediaList]

        if path not in self._watched:
            self._watched.append(path)
        self._owner.systemChanged = True

    @try_catch("Cannot list shared folder", [])
    @type_check
    def _listShared(self) -> list:
        """ List media files in a shared location.

        Returns:
            list -- List of media files.
        """

        path = self.sharedPath
        # path format should be: smb://username:password@host_ip/folder/
        if not path.startswith("smb://"):
            return []
        path = path[6:]

        if ":" not in path or "@" not in path:
            return []

        username = path.split(":")[0]
        path = path[len(username)+1:]

        password = "@".join(path.split("@")[:-1])
        path = path[len(password)+1:]

        server_ip = path.split("/")[0]
        path = path[len(server_ip)+1:]

        basepath = path.split("/")[0]
        path = path[len(basepath):]

        hostname = socket.gethostname()
        if hostname:
            hostname = hostname.split('.')[0]
        else:
            hostname = f"SMB{os.getpid()}"

        netBios = NetBIOS()
        server_name = netBios.queryIPForName(server_ip)[0]
        netBios.close()

        conn = SMBConnection(username, password, hostname,
                             server_name, use_ntlm_v2=True)
        assert conn.connect(server_ip)

        result = []
        subPaths = [path]  # recursion list
        while len(subPaths) > 0:
            for file in conn.listPath(basepath, subPaths[0]):
                if file.isDirectory and not file.filename.startswith("."):
                    subPaths.append(os.path.join(subPaths[0], file.filename))
                elif os.path.splitext(file.filename)[1] in self.SupportedFormats:
                    fullpath = self.sharedPath.strip(
                        "/") + os.path.join(subPaths[0], file.filename)
                    result.append(os.path.relpath(fullpath, self.sharedPath))
            subPaths.remove(subPaths[0])
        conn.close()
        return result

    @try_catch("Cannot convert subtitles")
    @type_check
    def _convertSubtitles(self, path: str) -> None:
        """ Convert subtitles to utf-8 encoding.

        Arguments:
            path {str} -- Path to media file.
        """

        root = os.path.splitext(path)[0]
        # convert encoding to utf8 without BOM
        if os.path.isfile(root + ".srt"):
            logger.debug("Convert subtitles to utf-8: %s.srt", root)
            try:
                with codecs.open(root + ".srt", "r", encoding="windows-1251") as f1:
                    content = f1.read()
                with codecs.open(root + ".srt", "w", encoding="utf8") as f2:
                    f2.write(content)
            except UnicodeDecodeError:
                pass

File: Systems/MediaPlayerSystem.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

using LibVLCSharp.Shared;

using MyHome.Utils;

using Newtonsoft.Json;

using NLog;

namespace MyHome.Systems
{
    public class MediaPlayerSystem : BaseSystem
    {
        // TODO: remove once verify agent's functionality
        private static readonly ILogger logger = LogManager.GetCurrentClassLogger();

        private static readonly string[] supportedFormats = {
            ".mkv", ".avi", ".mov", ".wmv", ".mp4", ".mpg", ".mpeg", ".m4v", ".3gp", ".mp3" };


        [UiProperty(true)]
        public List<string> MediaPaths { get; }

        [UiProperty(true)]
        public List<string> Radios { get; }

        [UiProperty(true, "MB")]
        public double SongsDiskUsage { get; set; } // MB


        private bool sortByDate;
        [UiProperty]
        public bool SortByDate
        {
            get => this.sortByDate;
            set
            {
                if (this.sortByDate != value)
                {
                    this.sortByDate = value;
                    this.RefreshMediaList();
                }
            }
        }

        [UiProperty]
        public int Volume { get; set; }

        [UiProperty]
        public List<string> Watched { get; }

        [JsonIgnore]
        public Dictionary<string, int> Songs => this.songsList.ToDictionary(s => s.Name, s => s.Rating); // song path / play count (rating)


        [JsonIgnore]
        [UiProperty]
        public List<string> MediaList => this.GetMediaList();

        private string playing;
        [JsonIgnore]
        [UiProperty]
        public string Playing
        {
            get
            {
                if (this.player.State == VLCState.Stopped)
                    this.playing = "";
                return this.playing;
            }
        }

        [JsonIgnore]
        [UiProperty]
        public bool Paused => this.player.State == VLCState.Paused;

        [JsonIgnore]
        [UiProperty]
        public string TimeDetails => this.GetTimeDetails();

        private sealed class SongInfo
        {
            public string Name { get; set; }
            public string Url { get; set; }
            public int Rating { get; set; }
        }
        [JsonProperty]
        private readonly List<SongInfo> songsList;
        private readonly List<string> mediaList;
        private DateTime lastRefreshMediaListTimer;
        private readonly LibVLC libVLC;
        private readonly MediaPlayer player;


        public MediaPlayerSystem()
        {
            this.MediaPaths = new List<string>();
            this.Radios = new List<string>();
            this.SongsDiskUsage = 500;
            this.sortByDate = false;
            this.Volume = 50;
            this.Watched = new List<string>();
            this.songsList = new List<SongInfo>();

            this.mediaList = new List<string>();
            this.lastRefreshMediaListTimer = DateTime.Now - TimeSpan.FromDays(1);
            this.libVLC = new LibVLC();
            this.player = new MediaPlayer(this.libVLC);
            this.playing = "";

            Directory.CreateDirectory(MyHome.Instance.Config.SongsPath);
        }

        public override void Stop()
        {
            base.Stop();
            this.StopMedia();
        }

        protected override void Update()
        {
            base.Update();

            // stop if end reached
            if (this.player.State == VLCState.Ended)
                this.player.Stop();
            System.Threading.Thread.Sleep(100);
        }


        public void Play(string path)
        {
            logger.Debug($"Play media: {path}");
            if (string.IsNullOrEmpty(path))
                return;

            this.playing = path;
            if (!path.StartsWith("http")) // if not URL remove media/radios prefix
                path = path[(path.IndexOf(":") + 1)..];

            this.player.Media = new Media(this.libVLC, new Uri(path));
            this.player.Media.AddOption(":subsdec-encoding=Windows-1251"); // set default encoding to Cyrillic
            this.player.Volume = this.Volume;
            this.player.Fullscreen = true;
            this.player.Play();
            this.MarkWatched(this.playing);
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaPlayed, this.playing);
        }

        public void StopMedia()
        {
            logger.Debug($"Stop media: {this.playing}");
            this.player.Stop();
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaStopped);
        }

        public void Pause()
        {
            logger.Debug($"Pause media: {this.playing}");
            this.player.Pause();
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaPaused);
        }

        public void VolumeDown()
        {
            logger.Debug($"Volume down media: {this.playing} to {this.Volume - 5}");
            this.Volume -= 5;
            this.player.Volume = this.Volume;
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaVolumeDown);
        }

        public void VolumeUp()
        {
            logger.Debug($"Volume up media: {this.playing} to {this.Volume + 5}");
            this.Volume += 5;
            this.player.Volume = this.Volume;
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaVolumeUp);
        }

        public void SeekBack()
        {
            logger.Debug($"Seek back media: {this.playing}");
            this.player.Time -= 30 * 1000; // -30 seconds
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaSeekBack);
        }

        public void SeekForward()
        {
            logger.Debug($"Seek forward media: {this.playing}");
            this.player.Time += 30 * 1000; // +30 seconds
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaSeekForward);
        }

        public void SeekBackFast()
        {
            logger.Debug($"Seek back fast media: {this.playing}");
            this.player.Time -= 600 * 1000; // -600 seconds
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaSeekBackFast);
        }

        public void SeekForwardFast()
        {
            logger.Debug($"Seek forward fast media: {this.playing}");
            this.player.Time += 600 * 1000; // +600 seconds
            MyHome.Instance.Events.Fire(this, GlobalEventTypes.MediaSeekForwardFast);
        }

        public void RefreshMediaList()
        {
            this.lastRefreshMediaListTimer = DateTime.Now - TimeSpan.FromDays(1);
        }

        public string AddSong(string url)
        {
            logger.Debug($"Add song: {url}");
            string name = url;
            if (url.Contains("youtube") || url.Contains("youtu.be"))
                name = Services.DownloadYouTubeAudioAsync(url, MyHome.Instance.Config.SongsPath).Result;
            if (!this.songsList.Any(s => s.Name == name))
                this.songsList.Add(new SongInfo() { Name = name, Url = url, Rating = this.songsList.Select(s => s.Rating).Max() });

            // cleanup songs if we exceed the usage capacity
            Utils.Utils.CleanupFilesByCapacity(this.songsList.OrderBy(s => s.Rating)
                .Select(s => Path.Join(MyHome.Instance.Config.SongsPath, s.Name))
                    .Where(p => File.Exists(p)).Select(p => new FileInfo(p)),
                this.SongsDiskUsage, logger);
            return name;
        }

        public string EnsureSong(string name)
        {
            if (string.IsNullOrEmpty(name) || File.Exists(Path.Join(MyHome.Instance.Config.SongsPath, name)))
                return name;

            var song = this.songsList.FirstOrDefault(s => s.Name == name);
            if (song == null)
                return name;
            if (song.Url.Contains("youtube") || song.Url.Contains("youtu.be"))
                song.Name = Services.DownloadYouTubeAudioAsync(song.Url, MyHome.Instance.Config.SongsPath).Result ?? song.Name;
            return song.Name;
        }

        public void IncreaseSongRating(string name, int rating = 1)
        {
            var song = this.songsList.FirstOrDefault(s => s.Name == name);
            if (song != null)
                song.Rating += rating;
        }


        private List<string> GetMediaList()
        {
            lock (this.mediaList)
            {
                if (DateTime.Now > this.lastRefreshMediaListTimer + TimeSpan.FromDays(1))
                {
                    logger.Debug("Refresh media list");
                    this.lastRefreshMediaListTimer = DateTime.Now;
                    this.mediaList.Clear();
                    for (int i = 0; i < this.MediaPaths.Count; i++)
                    {
                        if (!Directory.Exists(this.MediaPaths[i]))
                            continue;

                        var filePaths = Directory.EnumerateFiles(this.MediaPaths[i], "*.*", SearchOption.AllDirectories)
                            .Where(p => supportedFormats.Any(f => p.EndsWith(f)));
                        this.mediaList.AddRange(filePaths.Select(p => $"media{i + 1}:" + p));
                    }
                    if (this.sortByDate)
                    {
                        this.mediaList.Sort((a, b) => Directory.GetLastWriteTime(a[(a.IndexOf(":") + 1)..])
                            .CompareTo(Directory.GetLastWriteTime(b[(b.IndexOf(":") + 1)..])));
                    }
                    else
                        this.mediaList.Sort();
                    this.mediaList.AddRange(this.Radios.Select(r => "radios:" + r));
                }
            }
            return this.mediaList;
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0042:Deconstruct variable declaration")]
        private string GetTimeDetails()
        {
            static (long hours, long minutes, long seconds) GetTime(long millis)
            {
                var s = millis / 1000;
                var m = Math.DivRem(s, 60, out s);
                var h = Math.DivRem(m, 60, out m);
                return (hours: h, minutes: m, seconds: s);
            }

            var time = GetTime(this.player.Time);
            var length = GetTime(this.player.Length);
            return $"{time.hours:00}:{time.minutes:00} / {length.hours:00}:{length.minutes:00}";
        }

        private void MarkWatched(string path)
        {
            logger.Debug($"Mark as watched: {path}");

            if (!this.Watched.Contains(path))
                this.Watched.Add(path);

            // cleanup watched list from nonexistent files
            var list = this.GetMediaList();
            this.Watched.RemoveAll(w => !list.Contains(w));
            MyHome.Instance.SystemChanged = true;
        }
    }
}
